### PR TITLE
Ros2 crystal devel

### PR DESCRIPTION
**Pull request by Rohita83**
This PR adds following changes on crystal.
1. Fixed to build on crystal.
2. Fix failed few ekf/ukf testcases.
3. Automated scripts updated w.r.t. crystal.
4. Added new filter_base_diagnostics_timestamps testcases.
5. Update the README  "ros2_migration_readme.md"

Requesting to review for the same.

Thank you.

### CODE REVIEW

**Comment by ayrton04** [comment]
Looks like you have some conflicts to resolve here.

File: ros2_migration_readme.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+	1. System should have installed crystal distro. Check out installation instructions and tutorials https://index.ros.org/doc/ros2/.
+	2. User should checkout diagnostics pkg & build (Refer steps below).
+
+		Steps to checkout diagnostics pkg & build:
+
+			1. Open the new terminal and run below commands.
+			2. mkdir -p diagnostics_ws/src
+			3. cd ~/diagnostics_ws/src
+			4. git clone git@github.com:vaibhavbhadade/diagnostics.git //Note: branch should be remotes/origin/ros2-devel
```
Comment (ayrton04) [inline ros2_migration_readme.md]: You can roll this into the command itself:

```
git clone git@github.com:vaibhavbhadade/diagnostics.git -b ros2-devel
```

Comment (Rohita83) [reply ros2_migration_readme.md]: Agreed and fixed on https://github.com/cra-ros-pkg/robot_localization/pull/436/commits/ef2e6123bf5149739a013296292266f6073b3320

**Comment by Rohita83** [comment]
> Looks like you have some conflicts to resolve here.

I have ovservered that "Resolved conflicts" button is deactivated for me and I think conflicts will be resolve manually while merging into cra-ros-pkg:crystal-devel from Rohita83:ros2-crystal-devel.

**Comment by ayrton04** [comment]
You don't need to resolve the conflicts via the GitHub interface. The button is grayed out for me, too. You need to use the command line instructions.

**Comment by Rohita83** [comment]
> You don't need to resolve the conflicts via the GitHub interface. The button is grayed out for me, too. You need to use the command line instructions.

I guess, I don't have the access rights to resolve this conflict while merge on your repo (cra-ros-pkg:crystal-devel). Conflict would be resolve manually through command line when you will merge into cra-ros-pkg:crystal-devel from Rohita83:ros2-crystal-devel.

I have tried to merge it locally through command line as below:

```
user@PTL011729-1:~/GIT_ROS2/merge_issue/robot_localization$ git merge ros2-crystal-devel
Auto-merging test/test_ukf_localization_node_bag3.sh
CONFLICT (content): Merge conflict in test/test_ukf_localization_node_bag3.sh
Auto-merging test/test_ukf_localization_node_bag2.sh
Auto-merging test/test_ukf_localization_node_bag1.sh
Auto-merging test/test_ekf_localization_node_bag3.sh
CONFLICT (content): Merge conflict in test/test_ekf_localization_node_bag3.sh
Auto-merging test/test_ekf_localization_node_bag2.sh
CONFLICT (content): Merge conflict in test/test_ekf_localization_node_bag2.sh
Auto-merging test/test_ekf_localization_node_bag1.sh
Auto-merging test/robot_localization_test_cases.py
CONFLICT (modify/delete): test/README.md deleted in ros2-crystal-devel and modified in HEAD. Version HEAD of test/README.md left in tree.
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by ayrton04** [comment]
No, you can correct conflicts before I merge the change. It doesn't matter whether you have access to the main repo or not. In the error message you posted, the command line interface is telling you where the conflicts are. So you fix them, then run `git add <fixed files>`, then commit and push.

So open up `test/test_ukf_localization_node_bag3.sh`, and find where the conflict is. Then fix it, then `git add test/test_ukf_localization_node_bag3.sh`. Then repeat for the other files. The README might be tricky, since it was deleted.

**Comment by Rohita83** [comment]
@ayrton04 I have revert back the changes on https://github.com/cra-ros-pkg/robot_localization/pull/436/commits/df5f9852f1f39225ef2c4738c3c0a25e59c33c96 commit in below files to resolve conflicts now and i will make new pull request for these files changes w.r..t Crystal after current code get merged successfully.

```
1. test/test_ukf_localization_node_bag3.sh
2. test/test_ekf_localization_node_bag2.sh
3. test/test_ekf_localization_node_bag3.sh
4. test/README.md
```

**Review by ayrton04 — APPROVED** [review]

**Comment by ayrton04** [comment]
Thanks again! I plan on walking through this entire codebase soon, because r_l needs to be refactored (_badly_).

**Comment by Rohita83** [comment]
> Thanks again! I plan on walking through this entire codebase soon, because r_l needs to be refactored (_badly_).

Thanks @ayrton04 for accepting the robot_localization code changes for crystal. 

One thing that i have ovserved that testcases related to navsat_transform_node (output - sensor_msgs/NavSatFix message containing GPS data) not available for outdoor robots navigation using GPS data on ROS2 yet. Please let me know any suggesion on this.

Thank you.

**Comment by routiful** [comment]
@Rohita83 

Before I start my commets, I appreciated your work that port to ROS2 in this RL package.
I have tested se_node in crystal-devel with my sensors. The output is make sense.
But I can't understand how to implemented navsat_transform_node with my gps.

> One thing that i have ovserved that testcases related to navsat_transform_node (output - sensor_msgs/NavSatFix message containing GPS data) not available for outdoor robots navigation using GPS data on ROS2 yet. Please let me know any suggesion on this.

Is there any problem navsat_transform_node??

**Pull request merged**